server/src/main/java/org/elasticsearch/common/blobstore/BlobPath.java

```diff
@@ -26,16 +26,31 @@ public class BlobPath {
 
     private final List<String> paths;
 
+    private final Purpose purpose;
+
     private BlobPath(List<String> paths) {
+        this(paths, Purpose.GENERIC);
+    }
+
+    private BlobPath(List<String> paths, Purpose purpose) {
         this.paths = paths;
+        this.purpose = Objects.requireNonNull(purpose, "purpose must not be null");
     }
 
     public List<String> parts() {
         return paths;
     }
 
+    public Purpose purpose() {
+        return purpose;
+    }
+
     public BlobPath add(String path) {
-        return new BlobPath(CollectionUtils.appendToCopy(this.paths, path));
+        return new BlobPath(CollectionUtils.appendToCopy(this.paths, path), this.purpose);
+    }
+
+    public BlobPath purpose(Purpose purpose) {
+        return new BlobPath(List.copyOf(this.paths), purpose);
```

Comment:
Is it possible to change the purpose of current path?
If it is not the case, is it worth combining adding a path and setting the purpose into one call with something like `.add("/dir", purpose)` ?

Reply:
Since we are heading to a different direction, the change here is no longer applicable.